Overwrite old node labels for first node update to the endpoint manager (v1.14)

skmatti opened the pull request:
This fixes a race condition during agent startup where some node label updates were missed. If the k8s node watcher initializes before the host endpoint is created, the agent might miss some node label updates.

Fixes #29649

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 33551
```

pchaigno commented on pkg/endpoint/endpoint.go:
```diff
@@ -1927,7 +1927,12 @@ func (e *Endpoint) ModifyIdentityLabels(source string, addLabels, delLabels labe
 		return err
```
*(Not on a specific line of code on purpose.)*

I'd include more explanations in the commit description. Right now, your issue has a better explanation than the commit and I believe it would currently be hard for someone without context to understand your pull request. Reviewers are likely to struggle because of that.

One way to check this is to ask a colleague without context to review. If they fail to do so without reading the issue or asking you questions, then the commits probably don't have enough information.

skmatti replied:
Thanks, I just updated the commit description. WDYT?

squeed reviewed (changes requested):
This doesn't seem like the right way to fix this.

Rather, the issue is that we're passing the wrong information to `updateHostEndpointLabels()`.

My suggestion:

```go
// in ./pkg/endpointmanager/host.go
func (mgr *endpointManager) startNodeLabelsObserver(old map[string]string) {
	mgr.localNodeStore.Observe(context.Background(), func(ln node.LocalNode) {
		if maps.Equal(old, ln.Labels) {
			return
		}

		if err := mgr.updateHostEndpointLabels(old, ln.Labels); err != nil {
			log.Error(err)
		} else {
			old = ln.Labels
		}
	}, func(error) { /* Executed only when we are shutting down */ })
}

func (mgr *endpointManager) updateHostEndpointLabels(oldNodeLabels, newNodeLabels map[string]string) error {
	nodeEP := mgr.GetHostEndpoint()
	if nodeEP == nil {
		return fmt.Errorf("host endpoint not found")
	}

	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
	if err != nil {
		// An error can only occur if either the endpoint is terminating, or the
		// old labels are not found. Both are impossible, hence there's no point
		// in retrying.
		return fmt.Errorf("unable to update host endpoint labels: %w", err)
	}
	return nil
}
```

Otherwise we will still miss edges, since we update "old" without actually successfully applying. Make sense?

skmatti committed:
fix: set node labels correctly for first node update

This fixes a race condition during agent startup
where k8s node label updates are rejected
indefinitely by the host endpoint.

The current endpoint label update logic rejects
a request if any of the labels on the old node
are not present in the endpoint manager state.

For host endpoints, the k8s node label add/update
events may be missed if the k8s node watcher
intializes before the host endpoint is created.
As a result, the host endpoint labels are outdated,
and all subsequent node label updates are rejected
due to the aforementioned precondition check.

To resolve the issue, this commit sets the old labels
to nil on the first node update.

Also, collate the node watch handler into a single
function

Fixes #29649

Signed-off-by: Satish Matti <smatti@google.com>

squeed commented on pkg/endpointmanager/host.go:
```diff
@@ -36,20 +38,7 @@ func (mgr *endpointManager) HostEndpointExists() bool {
-	nodeEP.UpdateLabels(context.TODO(), labels.LabelSourceAny, newIdtyLabels, nil, false)
-
-	return nil
+	return mgr.OnUpdateNode(&slim_corev1.Node{}, newNode, swg)
```
nit: pass `nil` here, handle nil below. Otherwise, this looks a lot better.

skmatti replied:
Done

squeed commented on pkg/endpointmanager/host.go:
```diff
@@ -58,21 +47,31 @@ func (mgr *endpointManager) OnAddNode(newNode *slim_corev1.Node,
-	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
-	if err != nil {
+	if err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s); err != nil {
+		log.WithError(err).Error("Unable to update host endpoint labels")
```
Assuming this error message doesn't show up in regular lifecycle churn, I'm OK with this being an error. That said, the lifecycles of nodes and endpoints are theoretically independent -- I'm not sure what the invariants are here.

skmatti replied:
Yeah, this does not show up in regular lifecycle churn. This error message indicate a real failure if the node labels are not updated and host firewall could not applied correctly. The downstream does not log this error so it was difficult to debug this issue in our env.
